### PR TITLE
Clean up logs 3.x

### PR DESCRIPTION
**Pull request by iilyak**
## Overview

Port https://github.com/apache/couchdb/pull/3031 (Clean up logs) into 3.x branch.

This PR deals with cleaning up logs from user's data (potentially sensitive) in the cases when OTP processes crash. All commits except "Add format_status/2 callback in gen_server implementations" were cherry picked from https://github.com/apache/couchdb/pull/3031. The "Add format_status/2 callback in gen_server implementations" was extended to include applications relevant for 3.x.

## Testing recommendations

I tested it by running test suite and by comparing logs before and after. I did use following query to reduce noise.

```
find src/ -name couch.log | xargs -I '{}' cat '{}' | grep error | grep -v 'No Admin Account configured' | grep -v 'no_db_file' | grep -v 'no such file' | grep -v 'Request to create N=' | grep -v 'ddoc_cache_ev :: ' | grep -v '"error":"not_found"' | grep -v 'httpd 404' | grep -v boom | grep -v bam | grep -v forbidden | grep -v request_body_too | grep -v xxxxxxxxxxxxxxxxxxxxxxxxx | grep -v 'http 405' | grep -v method_not_allowed | grep -v db_not_found | grep -v 'source missing' | grep -v missing_source | grep -v 'list of tags' | grep -v 'Failed to set security object on' | grep -v 'fabric_rpc:group_info/2 exit:{killed,' | grep -v 'mem3_reshard db monitor' | grep -v unauthorized | grep -v too_large | grep -v 'httpd 413' | grep -v 'httpd 405' | grep -v not_acceptable | grep -v 'httpd [0-9]*' | grep -v 'unsupported encoding' | grep -v content_md5_mismatch | grep -v changes_req_failed | grep -v 'ignoring document with empty ID' | grep -v 'some_id <<"fail">>' | grep -v query_parse_error | grep -v 'Db was deleted' | grep -v kapow | grep -v bad_content_type | grep -v 'purge-test-stuff' | grep -v '{file,"src/couch_native_process.erl"},{line,258}' | grep -v 'file already exists' | grep -v 'Expression does not eval' | grep -v 'Transient job' | grep -v 'mem3_reshard_job cleaning' | grep -v 'mem3_reshard_job source' | grep -v 'Purge checkpoin' | grep -v 'missing_target' | grep -v 'target db missing'
```

I did run `make eunit` on vanilla 3.x branch and redirected the output of the above query into a file `3.x.vanilla.log`. Then I did remove all `couch.log` files and switched to `clean-up-logs-3.x` branch and repeated the process. The result is saved in `3.x.updated.log`. I did compare two files manually.

## Related Issues or Pull Requests

- https://github.com/apache/couchdb/pull/3031

## Checklist

- [x] Code is written and works correctly
- [ ] Changes are covered by tests
- [ ] Any new configurable parameters are documented in `rel/overlay/etc/default.ini`
- [ ] A PR for documentation changes has been made in https://github.com/apache/couchdb-documentation

### CODE REVIEW

File: src/chttpd/src/chttpd_node.erl [hunk 1]
```diff
@@ -71,7 +71,9 @@ handle_node_req(#httpd{method='PUT', path_parts=[_, Node, <<"_config">>, Section
     Value = couch_util:trim(chttpd:json_body(Req)),
     Persist = chttpd:header_value(Req, "X-Couch-Persist") /= "false",
     OldValue = call_node(Node, config, get, [Section, Key, ""]),
-    case call_node(Node, config, set, [Section, Key, ?b2l(Value), Persist]) of
+    IsSensitive = Section == <<"admins">>,
+    Opts = #{persisit => Persist, sensitive => IsSensitive},
```
Comment (rnewson) [inline src/chttpd/src/chttpd_node.erl:75]: typo

File: src/couch/src/couch_term.erl [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+format_pid(Pid) when is_pid(Pid) ->
+    pid_to_list(Pid);
+
+format_pid(Term) ->
+    format_term(Term).
+
+format_port(Port) when is_port(Port) ->
+    erlang:port_to_list(Port);
+
+format_port(Term) ->
```
Comment (rnewson) [inline src/couch/src/couch_term.erl:82]: why does format_port take non-port arguments? ditto for the other functions structured this way.

File: src/chttpd/src/chttpd_node.erl (hunk 1 above)
Comment (iilyak) [reply src/chttpd/src/chttpd_node.erl:75]: Good catch!!

**Review by rnewson — COMMENTED** [review]
I'm not sure this is appropriate for couchdb. I understand the motivation but it does not apply to all users/admins of couchdb and they should be allowed to keep the current behaviour where stack traces and process states are unaltered. This should be configurable.

I left some specific comments but this also seems like a very large diff for a conceptually much smaller change.

File: src/couch/src/couch_term.erl (hunk 2 above)
Comment (iilyak) [reply src/couch/src/couch_term.erl:82]: The main purpose for these functions is to format record fields. In some cases the value of a field can be `nil` or `undefined`. The default clause was added to simplify the caller side. Otherwise I would have to add a case for each filed I try to convert to each `format_status/2` function which uses this facility.

File: src/couch/src/couch_os_process.erl [hunk 3]
```diff
@@ -256,6 +256,24 @@ code_change(_, {os_proc, Cmd, Port, W, R, Timeout} , _) ->
...
+        idle = Idle
+    } = State,
+    MapState = #{
+        port => couch_term:format_port(Port),
+        writer => couch_term:format_fun(Writer),
+        reader => couch_term:format_fun(Reader),
+        timeout => TimeOut,
+        idle => Idle
+    },
+    [{data, [{"State", MapState}]}].
```
Comment (rnewson) [inline src/couch/src/couch_os_process.erl:275]: it is odd that we convert everything to maps for format_status even though that is not the format it takes really.

**Comment by iilyak** [comment]
>  I understand the motivation but it does not apply to all users/admins of couchdb and they should be allowed to keep the current behaviour where stack traces and process states are unaltered. 

I also tried to remove big terms from state. Do we want it to be configurable as well?

**Comment by iilyak** [comment]
> I'm not sure this is appropriate for couchdb. I understand the motivation but it does not apply to all users/admins of couchdb and they should be allowed to keep the current behaviour where stack traces and process states are unaltered. This should be configurable.

What do you think about the following approach. The format_status/2 would return a tuple `{Status :: term(), Options :: #{senstive => boolean(), records => %{name :: atom() => Fields :: list(atom())}}}`?

```
format_status(_Opt, [_PDict, State]) ->
    {[{data, [{"State", State}}]}], #{sensitive => false, records => #{
        state => record_info(fields, state),
        httpd => record_info(fields, httpd)
    }}}.
```

Then in couch_log application we would call user configured sanitizer module.

**Comment by rnewson** [comment]
I'd like to hear from non-IBM CouchDB committers that might have a different view on our scrubbing of stacktraces and process states and logs first. If it's generally desired then I'd agree it doesn't need to be configurable. My concern, with both my IBM hat and CouchDB hat on, is that we are degrading the quality of our diagnostics (albeit with good intention) and I think we need to proceed with a broader consensus.

I don't think format_status should return an erlang term that is completely different from the state of the process, it should look the same but with some items removed or redacted.

in summary, the goal is sound, we should ask if a solution is desired widely, whether it should be configurable and how closely the redacted output should match the output it replaces.

@janl @wohali @dch

File: src/couch/src/couch_term.erl (hunk 2 above)
Comment (iilyak) [reply src/couch/src/couch_term.erl:82]: Another option I came up with is this (the `format_port/2` can be expressed using `case` as well).

```
format_port(Term) ->
    format_port(Term, nil).
    
format_port(Term, _Default) when is_port(Term) ->
    erlang:port_to_list(Port);
    
format_port(_Term, Default) ->
    Default.    
```

**Comment by wohali** [comment]
If this is configurable and not enabled by default, I have no objection.

**Comment by iilyak** [comment]
Closing in favor of https://github.com/apache/couchdb/pull/3380
